FilenameBear.py: Enable default naming convention

yash-nisar opened the pull request:
Add default naming conventions that are widely accepted for
languages like java, javascript and python.

Closes https://github.com/coala/coala-bears/issues/1083

<!--
Thanks for your contribution!

Please take a quick look at those things down there. They're quite important.
Really! We wrote them for you. Yes you! With utmost care. Read them.
-->

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

yash-nisar commented:
@jayvdb Please have a look when you're free. :)

meetmangukiya commented:
This can be better implemented and maintained this way IMO:
- Create a dict that contains extensions as keys and lambdas that accept filename as argument and return the standard naming convention compliant name
- Then try 
```
try: 
  dict[extension](filename) 
except KeyError:
  # some nice message ?
```

This way there will be much less `if`s if we want to add other language supports in future :)

jayvdb commented on bears/general/FilenameBear.py:
```diff
@@ -37,8 +37,14 @@ def run(self, filename, file,
                 filename_without_extension)
+
+            if extension == '.js':
+                new_name = new_name.replace('_', '-')
```
the JS style should have its own proper name defined in the dict.

Create a separate issue & PR, as required, as this would be a feature.

jayvdb commented on bears/general/FilenameBear.py:
```diff
@@ -37,8 +37,14 @@ def run(self, filename, file,
         head, tail = os.path.split(filename)
         filename_without_extension, extension = os.path.splitext(tail)
         try:
+            if extension == '.java':
```
youch.  this is in a try block designed to catch the KeyError ... nothing else should be in here.

yash-nisar commented:
@Makman2 @javvdb Please have a look when you're free. :)

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -19,14 +20,20 @@ class FilenameBear(LocalBear):
+
     def run(self, filename, file,
-            file_naming_convention: str='snake',
+            file_naming_convention: str=None,
```
maybe make `auto` the default behaviour?

yash-nisar replied:
Okay :+1:

yash-nisar replied:
According to @jayvdb ,
> Ideally, we change the default file_naming_convention to None, but then we have two paths forward if the user doesnt set a value, 
> 1. use `snake`, but warn the user that they should be explicit, and allow auto for the user to explicitly request that we guess the right convention which will make it backwards compatible.

jayvdb replied:
Instead of `str`, I believe we have a LOV type specifier, which means you can remove the `.lower()` below

yash-nisar replied:
The tests were failing when I provided `Auto` instead of `auto` so I think we need to keep the `.lower()`. @jayvdb

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -19,14 +20,20 @@ class FilenameBear(LocalBear):
 
+    _language_naming_convention = {'.java': 'pascal',
+                                   '.js': 'kebab',
+                                   '.py': 'snake'}
```
I would allow it for this merge, but we should definitely introduce a simple system somewhere else mapping file endings to language identifiers, so we have a single interface for that kind of functionality :+1:

jayvdb replied:
https://github.com/AlejandroFrias/case-conversion might have one. If not, they might be a good maintainer of a simple mapping.
coala would eventually have a Language class approach

jayvdb replied:
This should be a module level private variable, and it will be extended, so the layout should be friendly to additions.
```py
x = {
  a,
  b,
}
```

jayvdb commented on bears/general/FilenameBear.py:
```diff
@@ -19,14 +19,20 @@ class FilenameBear(LocalBear):
 
         :param file_naming_convention:
             The naming-convention. Supported values are:
+            - ``auto`` to guess the right convention. Defaults to
```
s/right/correct/

yash-nisar commented:
@jayvdb @Makman2 Changes that I've performed since the last push:
* Added friendly layout for `_language_naming_convention = {
                                   '.java': 'pascal',
                                   '.js': 'kebab',
                                   '.py': 'snake',
                                  }`
* Changed `right` -> `correct` in docstring ` ``auto`` to guess the right convention...`
* Since the tests were failing when I removed the `.lower()` and supplied `Auto`, I chose to keep the `.lower()` or maybe we could open an issue for this in the core.

jayvdb commented:
> Since the tests were failing when I removed the .lower()

Did you find out why they failed?
Have you investigated replacing type annotation str with a list of values?

yash-nisar commented:
We haven't incorporated a conversion function yet IIUC because when I run 
`coala -V -b FilenameBear -S file_naming_convention=Snake -f xyz.py --flush-cache`, the value of the setting is `Snake` and when I run
`coala -V -b FilenameBear -S file_naming_convention=snake -f xyz.py --flush-cache`, the value of the setting is `snake` which I confirmed by using the debug mode.

yash-nisar commented:
Should we have this feature or should we expect the users to type in the exact setting ? @jayvdb

yash-nisar commented:
I've removed `.lower()` but this wont work if the user enters a setting as `AUTO` or `Auto` instead of `auto. @jayvdb

meetmangukiya commented on bears/general/FilenameBear.py:
```diff
@@ -19,14 +19,22 @@ class FilenameBear(LocalBear):
         :param file_naming_convention:
             The naming-convention. Supported values are:
+            - ``auto`` to guess the correct convention. Defaults to
+            ``snake`` if the correct convention cannot be guessed.
```
Align so that it starts just below the backticks of `auto` ?

yash-nisar replied:
I think it is alright since it is the continuation of the first point.

Makman2 replied:
hm right I think it improves readability in source when you align this to `auto` :+1:

yash-nisar replied:
Did not align because it caused indentation problems https://circleci.com/gh/coala/coala-bears/5890#tests/containers/1

meetmangukiya commented on bears/general/FilenameBear.py:
```diff
@@ -38,6 +46,21 @@ def run(self, filename, file,
         filename_without_extension, extension = os.path.splitext(tail)
+
+        if file_naming_convention is None:
+            self.err('Please specify a file naming convention explicitly. '
```
IMO, this is more of a warning, since we are falling back to something default. It should be an error if the bear can't run.

meetmangukiya commented on bears/general/FilenameBear.py:
```diff
@@ -38,6 +46,21 @@ def run(self, filename, file,
+
+        if file_naming_convention == 'auto':
+            if extension not in self._language_naming_convention.keys():
+                self.err('The file naming convention could not be guessed. '
```
Again, this should be warning, refer previous comments

yash-nisar commented:
Addressed your comments @meetmangukiya

yash-nisar commented:
@Makman2 Is it okay now ?

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -19,14 +19,22 @@ class FilenameBear(LocalBear):
+
     def run(self, filename, file,
-            file_naming_convention: str='snake',
+            file_naming_convention: str=None,
```
so what about making this `auto` by default? :)
(I think I commented that already...)

yash-nisar replied:
@jayvdb wanted a `None` option by default so we can warn the user to explicitly set the `file_naming_convention`

Makman2 replied:
alright :+1:

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -38,6 +46,21 @@ def run(self, filename, file,
+        if file_naming_convention is None:
+            self.warn('Please specify a file naming convention explicitly. '
+                      'Using the default "snake" naming convention...')
+            file_naming_convention = 'snake'
```
not `auto`?

yash-nisar replied:
The first point [here](https://github.com/coala/coala-bears/issues/1083#issuecomment-288518740) states that we use `snake` if the user doesn't mention explicitly. Is that okay ? @Makman2

Makman2 replied:
yeah though we should maybe use `auto` still, so the user directly gets the language-default. When you supply Java files and it defaults to snake you think "huh what the... why snake?".

Makman2 replied:
Oh and other point: I would add to the warning message:
```
Please specify a file naming convention explicitly or use "auto".
```
"explicitly specifying" would mean for me that you shouldn't use "auto", but instead `snake`, `camel` or others

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -38,6 +46,21 @@ def run(self, filename, file,
+            file_naming_convention = 'snake'
+
+        if file_naming_convention.lower() == 'auto':
+            if extension not in self._language_naming_convention.keys():
```
omit `.keys()`, as this is the same:
`extension not in self._language_naming_convention`

yash-nisar replied:
Done. :)

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -38,6 +46,21 @@ def run(self, filename, file,
+                file_naming_convention = 'snake'
+            else:
+                file_naming_convention = self._language_naming_convention[
+                    extension]
```
I would reverse order of the if, then you don't need to invert condition:
```
if extension in ...
    file_naming_convention = self._language_....
else:
    self.warn...
```

Makman2 replied:
you could also use `try-catch` with `KeyError`. Up to you :+1: :)

yash-nisar replied:
Roger that. :)

yash-nisar replied:
@Makman2 Should these naming conventions be case insensitive ?

yash-nisar replied:
Done. :)

Makman2 commented on tests/general/FilenameBearTest.py:
```diff
@@ -63,8 +62,58 @@ def test_space_case(self):
+
+    def test_auto_file_naming_convention_warning(self):
+        self.section['file_naming_convention'] = 'auto'
+        self.uut.execute('filename.cpp', [])
```
use some weirdo extension, cpp might be implemented :)

yash-nisar replied:
Alright @Makman2 , I'm on it now. :)

Makman2 commented on tests/general/FilenameBearTest.py:
```diff
@@ -6,7 +6,7 @@
 
 
-class SpaceConsistencyBearTest(LocalBearTestHelper):
+class FilenameBearTest(LocalBearTestHelper):
```
ohh damn copy paste virus :3
mind to put that into another commit? :)

yash-nisar replied:
Sure, why not ?As you say :D

yash-nisar commented:
To make things easier for reviewers, things that I've included since the last push: 
* Aligned `auto` with `snake`. See https://github.com/coala/coala-bears/pull/1530#discussion_r108040695
* Used weirdo extension of `filename.xyz`. See https://github.com/coala/coala-bears/pull/1530#discussion_r108041075
* Reversed the order of the `if` condition. See https://github.com/coala/coala-bears/pull/1530#discussion_r108040853
* Removed the `.keys()`. See https://github.com/coala/coala-bears/pull/1530#discussion_r108040835
* Added a separate commit for changing the classname.
* Added `.lower()` so `file_naming_convention` is case insensitive and added a test for the same. :smile: 
@Makman2 over to you. :)

yash-nisar commented:
@Makman2 cast that magic spell :P

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -19,14 +19,22 @@ class FilenameBear(LocalBear):
         :param file_naming_convention:
             The naming-convention. Supported values are:
+            - ``auto`` to guess the correct convention. Defaults to ``snake``
+            if the correct convention cannot be guessed.
```
alignment :3

yash-nisar replied:
Builds fail stating as `indentation problem` so I did not align it. Is it okay ? See https://github.com/coala/coala-bears/pull/1530#discussion_r108043976

Makman2 replied:
Ah okay nvm then :+1:

yash-nisar replied:
@Makman2 build failures due to indentation problems in yapf bear (completely unrelated to my PR). How should we handle this ?

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -38,8 +46,23 @@ def run(self, filename, file,
+
         try:
-            new_name = self._naming_convention[file_naming_convention](
+            new_name = self._naming_convention[file_naming_convention.lower()](
```
maybe you should lower `file_naming_convention` in the beginning in L54, then you don't need to lower it for comparison and here too. (You could do for example sth like this:
```
if file_naming_convention is None:
    self.warn...
    file_naming ... = auto
else:
    file_naming_convention = file_naming_convention.lower()
```
)

Makman2 commented on tests/general/FilenameBearTest.py:
```diff
@@ -46,7 +45,7 @@ def test_kebab_case(self):
 
     def test_pascal_case(self):
-        self.section['file_naming_convention'] = 'pascal'
+        self.section['file_naming_convention'] = 'Pascal'
```
up to you: you could test both, by wrapping into a loop:
```python
for convention in ('pascal', 'Pascal', 'PASCAL')
    self.section = ...
    self.check_validity...
```

Makman2 commented on tests/general/FilenameBearTest.py:
```diff
@@ -63,8 +62,58 @@ def test_space_case(self):
+        self.uut.execute('filename.xyz', [])
+
+        # Automatic bear messages we want to ignore.
+        self.uut.message_queue.get(timeout=1)
```
do you have to specify a timeout? I would invoke that without timeout, our tests do run max 40secs before pytest aborts the test ;)

Makman2 commented on bears/general/FilenameBear.py:
```diff
@@ -38,6 +46,23 @@ def run(self, filename, file,
+                    extension]
+            else:
+                self.warn('The file naming convention could not be guessed. '
+                          'Using the default "snake" naming convention...')
```
maybe turn the `...` into a single period `.`? Up to you :+1:

yash-nisar replied:
Alright :)

yash-nisar commented:
@Makman2 Can you ack so we can get this merged as soon as the issues that are blocking this get fixed ?

yash-nisar commented:
@Makman2 The tests pass now, one last review ?

Makman2 commented:
ack 1b1bdea 3c9c352

Makman2 commented:
@rultor merge

rultor commented:
> @rultor merge

@makman2 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/11290-290868651)

rultor commented:
> @rultor merge

@makman2 Done! FYI, the full log is [here](http://www.rultor.com/t/11290-290868651) (took me 1min)
